### ISSUE TITLE
Validation Constraint Inheritance Broken

### ISSUE THREAD

**Issue by scheb** [symfony/symfony#21567]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | >=3.1.9

Symfony validator does no longer inherit all constraints from the parent class. It seems like all the constraints from super-classes are lost when there are additional constraints on the same attribute in the sub-class. I was able to track it down to [this commit](https://github.com/symfony/symfony/commit/9dd6b0cf643380f3ffb67fcc85eef1255ecfb910), if I roll-back the changes it's working again. But couldn't wrap my head around yes, why this is happening.

Test script to reproduce the bug:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<constraint-mapping xmlns="http://symfony.com/schema/dic/constraint-mapping" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/constraint-mapping http://symfony.com/schema/dic/constraint-mapping/constraint-mapping-1.0.xsd">
    <class name="Foo">
        <property name="attribute">
            <constraint name="NotNull">
                <option name="message">NOT_NULL</option>
            </constraint>
        </property>
    </class>
    <class name="Bar">
        <property name="attribute">
            <constraint name="Type">
                <option name="type">string</option>
                <option name="message">NOSTRING</option>
            </constraint>
        </property>
    </class>
</constraint-mapping>
```

```php
<?php
include(__DIR__ . '/vendor/autoload.php');

use Symfony\Component\Validator\Validation;

abstract class Foo {
    public $attribute;
}

class Bar extends Foo {
}

$validator = Validation::createValidatorBuilder()->addXmlMapping(__DIR__ . '/validation.xml')->getValidator();

$test1 = new Bar();
$test1->attribute = null; // NotNull violation
var_dump($validator->validate($test1)->count()); // Should be 1, but is 0 in version >= 3.1.9

$test2 = new Bar();
$test2->attribute = 42; // Type violation
var_dump($validator->validate($test2)->count()); // Should be 1
```

**Comment by nicolas-grekas** [comment]
ping @uwej711

**Comment by scheb** [comment]
Just noticed, moving this piece of code
```php
// Include constraints from the parent class
$this->mergeConstraints($metadata);
```
from line 123 to line 115 right after the `$metadata = new ClassMetadata($class);` fixes it. That's the place where the logic was before the commit that I've mentioned.

**Comment by xabbuh** [comment]
I think this is caused by the changes made in #21053 (see also #21538). Can you confirm that?

**Comment by scheb** [comment]
Rolling back #21053 fixes it for me.

**Comment by xabbuh** [comment]
So it seems we have conflicting requirements here and in #15950 which are more or less mutually exclusive. I am actually not sure how to solve this sufficiently.

Maybe we can mitigate #15950 by modifying the patch from #21053 partially by skipping the merge only when constraints being used are the same (regardless their options)?

What do you think @symfony/deciders?

**Comment by ro0NL** [comment]
Imo. a sub class can only add additional constraints, at least by default that is. You're _extending_ by definition (`class Sub extends Base`).

That said.. what about `@Assert\Override()`?

**Comment by scheb** [comment]
Agree with @ro0NL that by default constraints should extend constraints from super-classes instead of overriding them.

In my opinion, it should only override property constraints, when the related property has been overwritten in the sub-class.  So the decision, which constraints should be applied, can only be made if we take into account the class hierarchy and how the properties of these classes are defined in addition to the constraint metadata.

To make clear what I mean, let's say we have constraints defined on both `Foo::$attribute` and `Bar::$attribute`.

"Extends" case:
```php
class Foo {
    public $attribute;
}

class Bar extends Foo {
}
```
- When validating a `Bar` object, `$attribute` constraints from both `Foo` and `Bar` should be applied.
- When validating a `Foo` object, `$attribute` constraints only from `Foo` should be applied.

"Override" case:
```php
class Foo {
    public $attribute;
}

class Bar extends Foo {
    public $attribute;
}
```
- When validating a `Bar` object, `$attribute` constraints only from `Bar` should be applied, because `Bar` overrides the attribute and therefore also the constraints from `Bar` should completely override the ones from `Foo`.
- When validating a `Foo` object, `$attribute` constraints only from `Foo` would be applied.

Overall it looks to me that the MetadataFactory needs to become aware of how the classes look like, it cannot just merge constraint metadata the one or the other way.

**Comment by ro0NL** [comment]
@scheb that's a way of looking at it :) but im not sure it's convenient to mix&match the PHP with the XML definition or so. Ie. adding the attribute in your "override" example changes the behavior of the XML definition.. that's weird.

For annotations it could work, as you need to define the property anyway to annotate it with constraints.. but imo. it's highly implicit behavior (ie. magic).

Basically we need a fix for #15950 only.

**Comment by przemyslaw-bogusz** [comment]
I have proposed `@Assert\EnableOverridingPropertyConstraints()` in #36155, which allows you to choose, if the constraints should be merged or overriden.
